File: onstock/internal/store/reports.go
package store

import (
	"fmt"
	"time"
)

// IncomeStatement es el Estado de Resultados (formato Honduras).
// Todos los montos de ventas son NETOS de ISV (el ISV no es ingreso de la empresa).
type IncomeStatement struct {
	From string `json:"from"`
	To   string `json:"to"`

	VentasBrutas float64 `json:"ventas_brutas"`
	Descuentos   float64 `json:"descuentos"`
	VentasNetas  float64 `json:"ventas_netas"`

	CostoVentas   float64 `json:"costo_ventas"`
	UtilidadBruta float64 `json:"utilidad_bruta"`

	GastosVentas         float64 `json:"gastos_ventas"`
	GastosAdministrativos float64 `json:"gastos_administrativos"`
	GastosOperativos     float64 `json:"gastos_operativos"`
	UtilidadOperativa    float64 `json:"utilidad_operativa"`

	GastosFinancieros float64 `json:"gastos_financieros"`
	OtrosGastos       float64 `json:"otros_gastos"`

	UtilidadAntesISR float64 `json:"utilidad_antes_isr"`
	ISRRate          float64 `json:"isr_rate"`
	ISR              float64 `json:"isr"`
	UtilidadNeta     float64 `json:"utilidad_neta"`

	ISVCobrado  float64 `json:"isv_cobrado"`
	NumVentas   int     `json:"num_ventas"`
	MargenBruto float64 `json:"margen_bruto"`
	MargenNeto  float64 `json:"margen_neto"`
}

func (s *Store) IncomeStatement(from, to string) (IncomeStatement, error) {
	st := IncomeStatement{From: from, To: to, ISRRate: s.settingFloat("isr_rate", 25)}

	err := s.db.QueryRow(`SELECT
	    COALESCE(SUM(subtotal + discount_net),0),
	    COALESCE(SUM(discount_net),0),
	    COALESCE(SUM(subtotal),0),
	    COALESCE(SUM(cost_total),0),
	    COALESCE(SUM(isv),0),
	    COUNT(*)
	  FROM sales WHERE status='completada' AND date(sale_date) BETWEEN ? AND ?`, from, to).
		Scan(&st.VentasBrutas, &st.Descuentos, &st.VentasNetas, &st.CostoVentas, &st.ISVCobrado, &st.NumVentas)
	if err != nil {
		return st, err
	}

	rows, err := s.db.Query(`SELECT category, COALESCE(SUM(amount),0) FROM expenses
	  WHERE expense_date BETWEEN ? AND ? GROUP BY category`, from, to)
	if err != nil {
		return st, err
	}
	defer rows.Close()
	for rows.Next() {
		var cat string
		var amt float64
		if err := rows.Scan(&cat, &amt); err != nil {
			return st, err
		}
		switch cat {
		case "ventas":
			st.GastosVentas = amt
		case "administrativos":
			st.GastosAdministrativos = amt
		case "financieros":
			st.GastosFinancieros = amt
		default:
			st.OtrosGastos += amt
		}
	}
	if err := rows.Err(); err != nil {
		return st, err
	}

	st.UtilidadBruta = st.VentasNetas - st.CostoVentas
	st.GastosOperativos = st.GastosVentas + st.GastosAdministrativos
	st.UtilidadOperativa = st.UtilidadBruta - st.GastosOperativos
	st.UtilidadAntesISR = st.UtilidadOperativa - st.GastosFinancieros - st.OtrosGastos
	if st.UtilidadAntesISR > 0 {
		st.ISR = round2(st.UtilidadAntesISR * st.ISRRate / 100)
	}
	st.UtilidadNeta = round2(st.UtilidadAntesISR - st.ISR)
	if st.VentasNetas > 0 {
		st.MargenBruto = round2(st.UtilidadBruta / st.VentasNetas * 100)
		st.MargenNeto = round2(st.UtilidadNeta / st.VentasNetas * 100)
	}
	for _, p := range []*float64{&st.VentasBrutas, &st.Descuentos, &st.VentasNetas, &st.CostoVentas,
		&st.UtilidadBruta, &st.GastosOperativos, &st.UtilidadOperativa, &st.UtilidadAntesISR, &st.ISVCobrado} {
		*p = round2(*p)
	}
	return st, nil
}

type TopProduct struct {
	ProductID int64   `json:"product_id"`
	Name      string  `json:"name"`
	SKU       string  `json:"sku"`
	Qty       float64 `json:"qty"`
	Revenue   float64 `json:"revenue"` // neto de ISV
	Profit    float64 `json:"profit"`
}

func (s *Store) TopProducts(from, to string, limit int) ([]TopProduct, error) {
	if limit <= 0 {
		limit = 10
	}
	rows, err := s.db.Query(`SELECT i.product_id, p.name, p.sku,
	    SUM(i.qty), SUM(i.qty*i.unit_price), SUM(i.qty*(i.unit_price-i.unit_cost))
	  FROM sale_items i
	  JOIN sales v ON v.id=i.sale_id AND v.status='completada' AND date(v.sale_date) BETWEEN ? AND ?
	  JOIN products p ON p.id=i.product_id
	  GROUP BY i.product_id ORDER BY SUM(i.qty*i.unit_price) DESC LIMIT ?`, from, to, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []TopProduct{}
	for rows.Next() {
		var t TopProduct
		if err := rows.Scan(&t.ProductID, &t.Name, &t.SKU, &t.Qty, &t.Revenue, &t.Profit); err != nil {
			return nil, err
		}
		t.Revenue = round2(t.Revenue)
		t.Profit = round2(t.Profit)
		out = append(out, t)
	}
	return out, rows.Err()
}

type MonthPoint struct {
	Month   string  `json:"month"` // YYYY-MM
	Label   string  `json:"label"` // ene 25
	Ventas  float64 `json:"ventas"`
	Utilidad float64 `json:"utilidad"`
}

var monthNamesES = [...]string{"ene", "feb", "mar", "abr", "may", "jun", "jul", "ago", "sep", "oct", "nov", "dic"}

// SalesSeries devuelve ventas netas y utilidad bruta por mes para los últimos n meses.
func (s *Store) SalesSeries(n int) ([]MonthPoint, error) {
	now := time.Now()
	points := make([]MonthPoint, 0, n)
	byMonth := map[string]*MonthPoint{}
	for i := n - 1; i >= 0; i-- {
		t := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.Local).AddDate(0, -i, 0)
		key := t.Format("2006-01")
		points = append(points, MonthPoint{Month: key, Label: fmt.Sprintf("%s %02d", monthNamesES[t.Month()-1], t.Year()%100)})
		byMonth[key] = &points[len(points)-1]
	}
	start := points[0].Month + "-01"
	rows, err := s.db.Query(`SELECT strftime('%Y-%m', sale_date), COALESCE(SUM(subtotal),0), COALESCE(SUM(subtotal-cost_total),0)
	  FROM sales WHERE status='completada' AND date(sale_date) >= ?
	  GROUP BY strftime('%Y-%m', sale_date)`, start)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	for rows.Next() {
		var key string
		var v, u float64
		if err := rows.Scan(&key, &v, &u); err != nil {
			return nil, err
		}
		if p, ok := byMonth[key]; ok {
			p.Ventas = round2(v)
			p.Utilidad = round2(u)
		}
	}
	return points, rows.Err()
}

type Dashboard struct {
	Month           string       `json:"month"`
	VentasMes       float64      `json:"ventas_mes"`
	UtilidadBrutaMes float64     `json:"utilidad_bruta_mes"`
	GastosMes       float64      `json:"gastos_mes"`
	NumVentasMes    int          `json:"num_ventas_mes"`
	TicketPromedio  float64      `json:"ticket_promedio"`
	ISVCobradoMes   float64      `json:"isv_cobrado_mes"`
	ValorInventario float64      `json:"valor_inventario"`
	ProductosActivos int         `json:"productos_activos"`
	LowStockCount   int          `json:"low_stock_count"`
	Series          []MonthPoint `json:"series"`
	TopProducts     []TopProduct `json:"top_products"`
	LowStock        []Product    `json:"low_stock"`
	RecentSales     []Sale       `json:"recent_sales"`
}

func (s *Store) Dashboard() (Dashboard, error) {
	now := time.Now()
	first := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.Local).Format("2006-01-02")
	today := now.Format("2006-01-02")
	d := Dashboard{Month: now.Format("2006-01")}

	err := s.db.QueryRow(`SELECT COALESCE(SUM(subtotal),0), COALESCE(SUM(subtotal-cost_total),0),
	    COALESCE(SUM(isv),0), COUNT(*)
	  FROM sales WHERE status='completada' AND date(sale_date) BETWEEN ? AND ?`, first, today).
		Scan(&d.VentasMes, &d.UtilidadBrutaMes, &d.ISVCobradoMes, &d.NumVentasMes)
	if err != nil {
		return d, err
	}
	if d.NumVentasMes > 0 {
		d.TicketPromedio = round2(d.VentasMes / float64(d.NumVentasMes))
	}
	d.VentasMes = round2(d.VentasMes)
	d.UtilidadBrutaMes = round2(d.UtilidadBrutaMes)
	d.ISVCobradoMes = round2(d.ISVCobradoMes)

	if err := s.db.QueryRow(`SELECT COALESCE(SUM(amount),0) FROM expenses WHERE expense_date BETWEEN ? AND ?`, first, today).Scan(&d.GastosMes); err != nil {
		return d, err
	}
	d.GastosMes = round2(d.GastosMes)

	if d.ValorInventario, err = s.InventoryValue(); err != nil {
		return d, err
	}
	if err := s.db.QueryRow(`SELECT COUNT(*) FROM products WHERE active=1`).Scan(&d.ProductosActivos); err != nil {
		return d, err
	}
	if err := s.db.QueryRow(`SELECT COUNT(*) FROM products WHERE active=1 AND stock <= min_stock`).Scan(&d.LowStockCount); err != nil {
		return d, err
	}
	if d.Series, err = s.SalesSeries(12); err != nil {
		return d, err
	}
	if d.TopProducts, err = s.TopProducts(first, today, 5); err != nil {
		return d, err
	}
	if d.LowStock, err = s.ListProducts(ProductFilter{LowStock: true}); err != nil {
		return d, err
	}
	if len(d.LowStock) > 8 {
		d.LowStock = d.LowStock[:8]
	}
	if d.RecentSales, err = s.ListSales(SaleFilter{Limit: 8}); err != nil {
		return d, err
	}
	return d, nil
}

// MonthlySummary agrupa todo lo necesario para el "resumen del mes".
type MonthlySummary struct {
	Year             int             `json:"year"`
	Month            int             `json:"month"`
	Statement        IncomeStatement `json:"statement"`
	TicketPromedio   float64         `json:"ticket_promedio"`
	ComprasRecibidas float64         `json:"compras_recibidas"`
	NumCompras       int             `json:"num_compras"`
	TopProducts      []TopProduct    `json:"top_products"`
	ValorInventario  float64         `json:"valor_inventario"`
}

func (s *Store) MonthlySummary(year, month int) (MonthlySummary, error) {
	first := time.Date(year, time.Month(month), 1, 0, 0, 0, 0, time.Local)
	last := first.AddDate(0, 1, -1)
	from, to := first.Format("2006-01-02"), last.Format("2006-01-02")

	ms := MonthlySummary{Year: year, Month: month}
	var err error
	if ms.Statement, err = s.IncomeStatement(from, to); err != nil {
		return ms, err
	}
	if ms.Statement.NumVentas > 0 {
		ms.TicketPromedio = round2(ms.Statement.VentasNetas / float64(ms.Statement.NumVentas))
	}
	err = s.db.QueryRow(`SELECT COUNT(*), COALESCE(SUM((SELECT SUM(qty*unit_cost) FROM purchase_order_items WHERE po_id=o.id)),0)
	  FROM purchase_orders o WHERE o.status='recibida' AND o.received_date BETWEEN ? AND ?`, from, to).
		Scan(&ms.NumCompras, &ms.ComprasRecibidas)
	if err != nil {
		return ms, err
	}
	ms.ComprasRecibidas = round2(ms.ComprasRecibidas)
	if ms.TopProducts, err = s.TopProducts(from, to, 10); err != nil {
		return ms, err
	}
	if ms.ValorInventario, err = s.InventoryValue(); err != nil {
		return ms, err
	}
	return ms, nil
}

// SalesReportRows: filas para exportar ventas de un período.
func (s *Store) SalesReportRows(from, to string) ([]Sale, error) {
	return s.ListSales(SaleFilter{From: from, To: to, Limit: 100000})
}

File: onstock/server/store/reports.js
// Traducción de internal/store/reports.go.
//
// Estado de Resultados en formato Honduras: todos los montos de ventas son
// NETOS de ISV (el ISV no es ingreso de la empresa). Los puntos de redondeo van
// exactamente donde los puso el original — mover uno solo cambia el cuadre.
import { round2 } from '../db.js';
import { listProducts } from './products.js';
import { listSales } from './sales.js';
import { inventoryValue } from './inventory.js';
import { ymd, ym, primerDiaDelMes, addDate, MESES_CORTO_ES } from '../lib/fechas.js';

export function incomeStatement(db, from, to) {
  const st = {
    from,
    to,
    ventas_brutas: 0,
    descuentos: 0,
    ventas_netas: 0,
    costo_ventas: 0,
    utilidad_bruta: 0,
    gastos_ventas: 0,
    gastos_administrativos: 0,
    gastos_operativos: 0,
    utilidad_operativa: 0,
    gastos_financieros: 0,
    otros_gastos: 0,
    utilidad_antes_isr: 0,
    isr_rate: db.settingFloat('isr_rate', 25),
    isr: 0,
    utilidad_neta: 0,
    isv_cobrado: 0,
    num_ventas: 0,
    margen_bruto: 0,
    margen_neto: 0,
  };

  const v = db.get(
    `SELECT
	    COALESCE(SUM(subtotal + discount_net),0) AS brutas,
	    COALESCE(SUM(discount_net),0)            AS descuentos,
	    COALESCE(SUM(subtotal),0)                AS netas,
	    COALESCE(SUM(cost_total),0)              AS costo,
	    COALESCE(SUM(isv),0)                     AS isv,
	    COUNT(*)                                 AS n
	  FROM sales WHERE status='completada' AND date(sale_date) BETWEEN ? AND ?`,
    from, to,
  );
  st.ventas_brutas = v.brutas;
  st.descuentos = v.descuentos;
  st.ventas_netas = v.netas;
  st.costo_ventas = v.costo;
  st.isv_cobrado = v.isv;
  st.num_ventas = v.n;

  const gastos = db.all(
    `SELECT category, COALESCE(SUM(amount),0) AS amt FROM expenses
	  WHERE expense_date BETWEEN ? AND ? GROUP BY category`,
    from, to,
  );
  for (const g of gastos) {
    switch (g.category) {
      case 'ventas': st.gastos_ventas = g.amt; break;
      case 'administrativos': st.gastos_administrativos = g.amt; break;
      case 'financieros': st.gastos_financieros = g.amt; break;
      default: st.otros_gastos += g.amt; break;
    }
  }
  // Se redondea antes de derivar los totales: SUM() sobre montos con dos
  // decimales arrastra ruido binario y ese ruido terminaba escrito tal cual en
  // la celda del Excel (-869.5699999999999 en la barra de fórmulas).
  st.gastos_ventas = round2(st.gastos_ventas);
  st.gastos_administrativos = round2(st.gastos_administrativos);
  st.gastos_financieros = round2(st.gastos_financieros);
  st.otros_gastos = round2(st.otros_gastos);

  st.utilidad_bruta = st.ventas_netas - st.costo_ventas;
  st.gastos_operativos = st.gastos_ventas + st.gastos_administrativos;
  st.utilidad_operativa = st.utilidad_bruta - st.gastos_operativos;
  st.utilidad_antes_isr = st.utilidad_operativa - st.gastos_financieros - st.otros_gastos;
  if (st.utilidad_antes_isr > 0) {
    st.isr = round2((st.utilidad_antes_isr * st.isr_rate) / 100);
  }
  st.utilidad_neta = round2(st.utilidad_antes_isr - st.isr);
  if (st.ventas_netas > 0) {
    st.margen_bruto = round2((st.utilidad_bruta / st.ventas_netas) * 100);
    st.margen_neto = round2((st.utilidad_neta / st.ventas_netas) * 100);
  }
  for (const k of ['ventas_brutas', 'descuentos', 'ventas_netas', 'costo_ventas',
    'utilidad_bruta', 'gastos_operativos', 'utilidad_operativa', 'utilidad_antes_isr', 'isv_cobrado']) {
    st[k] = round2(st[k]);
  }
  return st;
}

export function topProducts(db, from, to, limit) {
  if (!limit || limit <= 0) limit = 10;
  // sale_items.unit_price es el precio neto ANTES del descuento de la venta y
  // viene redondeado a dos decimales, así que sumar las líneas nunca da el
  // mismo número que sales.subtotal. Para que este reporte cuadre exactamente
  // con VentasNetas/UtilidadBruta del Estado de Resultados, repartimos el
  // subtotal real de cada venta entre sus líneas en proporción a lo que pesa
  // cada una: k = subtotal / SUM(qty*unit_price).
  const rows = db.all(
    `SELECT i.product_id, p.name, p.sku,
	    SUM(i.qty) AS qty,
	    SUM(i.qty*i.unit_price*f.k) AS revenue,
	    SUM(i.qty*i.unit_price*f.k - i.qty*i.unit_cost) AS profit
	  FROM sale_items i
	  JOIN sales v ON v.id=i.sale_id AND v.status='completada' AND date(v.sale_date) BETWEEN ? AND ?
	  JOIN products p ON p.id=i.product_id
	  JOIN (SELECT s2.id AS sid,
	          COALESCE(s2.subtotal/NULLIF(SUM(i2.qty*i2.unit_price),0),1) AS k
	        FROM sales s2 JOIN sale_items i2 ON i2.sale_id=s2.id
	        WHERE s2.status='completada' AND date(s2.sale_date) BETWEEN ? AND ?
	        GROUP BY s2.id) f ON f.sid=v.id
	  GROUP BY i.product_id ORDER BY 5 DESC LIMIT ?`,
    from, to, from, to, limit,
  );
  return rows.map((t) => ({
    product_id: t.product_id,
    name: t.name,
    sku: t.sku,
    qty: t.qty,
    revenue: round2(t.revenue),
    profit: round2(t.profit),
  }));
}

// salesSeries devuelve ventas netas y utilidad bruta por mes para los últimos n meses.
export function salesSeries(db, n) {
  const now = new Date();
  const points = [];
  const byMonth = new Map();
  for (let i = n - 1; i >= 0; i--) {
    const t = addDate(primerDiaDelMes(now), 0, -i, 0);
    const key = ym(t);
    const p = {
      month: key,
      label: `${MESES_CORTO_ES[t.getMonth()]} ${String(t.getFullYear() % 100).padStart(2, '0')}`,
      ventas: 0,
      utilidad: 0,
    };
    points.push(p);
    byMonth.set(key, p);
  }
  const start = `${points[0].month}-01`;
  const rows = db.all(
    `SELECT strftime('%Y-%m', sale_date) AS k, COALESCE(SUM(subtotal),0) AS v, COALESCE(SUM(subtotal-cost_total),0) AS u
	  FROM sales WHERE status='completada' AND date(sale_date) >= ?
	  GROUP BY strftime('%Y-%m', sale_date)`,
    start,
  );
  for (const r of rows) {
    const p = byMonth.get(r.k);
    if (p) {
      p.ventas = round2(r.v);
      p.utilidad = round2(r.u);
    }
  }
  return points;
}

export function dashboard(db) {
  const now = new Date();
  const first = ymd(primerDiaDelMes(now));
  const today = ymd(now);
  const d = {
    month: ym(now),
    ventas_mes: 0,
    utilidad_bruta_mes: 0,
    gastos_mes: 0,
    num_ventas_mes: 0,
    ticket_promedio: 0,
    isv_cobrado_mes: 0,
    valor_inventario: 0,
    productos_activos: 0,
    low_stock_count: 0,
    series: [],
    top_products: [],
    low_stock: [],
    recent_sales: [],
  };

  const v = db.get(
    `SELECT COALESCE(SUM(subtotal),0) AS ventas, COALESCE(SUM(subtotal-cost_total),0) AS utilidad,
	    COALESCE(SUM(isv),0) AS isv, COUNT(*) AS n
	  FROM sales WHERE status='completada' AND date(sale_date) BETWEEN ? AND ?`,
    first, today,
  );
  d.ventas_mes = v.ventas;
  d.utilidad_bruta_mes = v.utilidad;
  d.isv_cobrado_mes = v.isv;
  d.num_ventas_mes = v.n;
  if (d.num_ventas_mes > 0) d.ticket_promedio = round2(d.ventas_mes / d.num_ventas_mes);
  d.ventas_mes = round2(d.ventas_mes);
  d.utilidad_bruta_mes = round2(d.utilidad_bruta_mes);
  d.isv_cobrado_mes = round2(d.isv_cobrado_mes);

  d.gastos_mes = round2(db.scalar(
    'SELECT COALESCE(SUM(amount),0) FROM expenses WHERE expense_date BETWEEN ? AND ?', first, today,
  ));

  d.valor_inventario = inventoryValue(db);
  d.productos_activos = db.scalar('SELECT COUNT(*) FROM products WHERE active=1');
  d.low_stock_count = db.scalar('SELECT COUNT(*) FROM products WHERE active=1 AND stock <= min_stock');
  d.series = salesSeries(db, 12);
  d.top_products = topProducts(db, first, today, 5);
  d.low_stock = listProducts(db, { lowStock: true });
  if (d.low_stock.length > 8) d.low_stock = d.low_stock.slice(0, 8);
  d.recent_sales = listSales(db, { limit: 8 });
  return d;
}

// monthlySummary agrupa todo lo necesario para el "resumen del mes".
export function monthlySummary(db, year, month) {
  const first = new Date(year, month - 1, 1);
  const last = addDate(first, 0, 1, -1);
  const from = ymd(first);
  const to = ymd(last);

  const ms = {
    year,
    month,
    statement: incomeStatement(db, from, to),
    ticket_promedio: 0,
    compras_recibidas: 0,
    num_compras: 0,
    top_products: [],
    valor_inventario: 0,
  };
  if (ms.statement.num_ventas > 0) {
    ms.ticket_promedio = round2(ms.statement.ventas_netas / ms.statement.num_ventas);
  }
  const c = db.get(
    `SELECT COUNT(*) AS n, COALESCE(SUM((SELECT SUM(qty*unit_cost) FROM purchase_order_items WHERE po_id=o.id)),0) AS total
	  FROM purchase_orders o WHERE o.status='recibida' AND o.received_date BETWEEN ? AND ?`,
    from, to,
  );
  ms.num_compras = c.n;
  ms.compras_recibidas = round2(c.total);
  ms.top_products = topProducts(db, from, to, 10);
  ms.valor_inventario = inventoryValue(db);
  return ms;
}

// salesReportRows: filas para exportar ventas de un período.
export function salesReportRows(db, from, to) {
  return listSales(db, { from, to, limit: 100000 });
}
